arduino/smartcar/smartcar.ino

```diff
@@ -66,273 +72,245 @@ std::vector<char> frameBuffer;
+    frameBuffer.resize(Camera.width() * Camera.height() * Camera.bytesPerPixel());
+    mqtt.begin("aerostun.dev", 1883, WiFi);
+    #else
+    mqtt.begin(net);
+    #endif
+    if (mqtt.connect("arduino", "public", "public")) {
+        mqtt.subscribe("/smartcar/group16/#", 1);
+        mqtt.onMessage(+[](String& topic, String& message) {
+        if (topic == "/smartcar/group16/control/throttle") {
+            currentSpeed = message.toInt();
+            car.setSpeed(currentSpeed);
+        }
+        else if (topic == "/smartcar/group16/control/steering") {
+            car.setAngle(message.toInt());
+        }
+        else if (topic == "/smartcar/group16/auto/size") {
+            area = message.toInt() * 100;
+        }
+        else if (topic=="/smartcar/group16/auto/speed"){
+            velocity = message.toInt();
+        }
+        else if (topic == "/smartcar/group16/auto/pattern"){
+            pattern = message.toInt();
+        }
+        else {
+            Serial.println(topic + " " + message);
+        }
     });
   }
 }
 
+
 void loop() {
     obstacleAvoidance();
     handlePatterns();
-
     if (mqtt.connected()) {
-    mqtt.loop();
-
-     const auto currentTime = millis();
-#ifdef __SMCE__
-    static auto previousFrame = 0UL;
-    if (currentTime - previousFrame >= 65) {
-      previousFrame = currentTime;
-      Camera.readFrame(frameBuffer.data());
-      mqtt.publish("/smartcar/group16/camera", frameBuffer.data(), frameBuffer.size(),
-                   false, 0);
-    }
-#endif
-    static auto previousTransmission = 0UL;
-    if (currentTime - previousTransmission >= oneSecond)
-    {
-      previousTransmission = currentTime;
-      const auto distance = String(front.getDistance());
-      mqtt.publish("/smartcar/ultrasound/front", String(distance));
-      mqtt.publish("/smartcar/group16/speed", String(car.getSpeed()));
-      mqtt.publish("/smartcar/group16/distance", String(distanceInMeter()));
-      mqtt.publish("/smartcar/group16/obstacleMsg", String(obstacleDetectionMessage()));
-      mqtt.publish("/smartcar/group16/bagfull",String(bagFilledProgress()));
-      mqtt.publish("/smartcar/group16/completion", String(isCompleted));
+        mqtt.loop();
+        const auto currentTime = millis();
+        #ifdef __SMCE__
+        static auto previousFrame = 0UL;
+        if (currentTime - previousFrame >= 65) {
+            previousFrame = currentTime;
+            Camera.readFrame(frameBuffer.data());
+            mqtt.publish("/smartcar/group16/camera", frameBuffer.data(), frameBuffer.size(),false, 0);
+        }
+        #endif
+        static auto previousTransmission = 0UL;
+        if (currentTime - previousTransmission >= oneSecond){
+            previousTransmission = currentTime;
+            const auto distance = String(front.getDistance());
+            mqtt.publish("/smartcar/ultrasound/front", String(distance));
+            mqtt.publish("/smartcar/group16/speed", String(car.getSpeed()));
```

Comment:
We need to send the speedInPrecentege here instead of car.getSpeed

Reply:
Can you test  speedInPrecentege() since my app is still showing float number?

Reply:
yes!